cilium/CFP-32759-cilium-mtls-via-mtls-agent.md

```diff
@@ -0,0 +1,304 @@
+* An eBPF component that transparently redirects relevant traffic to the mTLS agent
+* The existing Cilium agent, which provides configuration to the mTLS agent and shares relevant workload information with the eBPF component required for redirection
+* Support for network policy enforcement. See [Policy Enforcement](#policy-enforcement-in-cilium-mtls) in Open Questions
+
+The following diagram highlights the key components in the mTLS tunneling workflow
+
+![alt text](images/cilium-mTLS-basic-workflow.png)
+
+### Components
+
+#### mTLS Agent
+
+The mTLS agent facilitates workload identity based mTLS connections between node agents that are transparent to the workloads making and receiving requests. It runs separately from the Cilium agent as a part of its own daemonset. Whether or not the mTLS agent will be running its own separate network namespace or the root network namespace is an open question. See Should the mTLS Agent be in the root network namespace or a separate network namespace? for more discussion on the implications of the mTLS agents placement. The Datapath example assumes the mTLS agent is in its own network namespace.
+
+To support the pluggable and interoperable mTLS agent model, the proxying node agent must adhere to the following specifications:
+* Support receiving and requesting configuration via the xDS protocol. Not required to implement the entire [xds API](https://github.com/cncf/xds)
+* Support transparent proxying (i.e., forwards the request without the original client or server knowing)
+* Mutually validate workload identity using mTLS 
+* Ability to request workload certificates
+* Support identity-based policy enforcement using the certificates used for mTLS encryption
+* Support the HTTP Connect protocol with specific port and header conventions for agent to agent communication (see [Protocol](#protocol) for information about HTTP Connect tunneling)
+
+One potential implementation of the mTLS agent is the [ztunnel](https://github.com/istio/ztunnel) L4 node proxy. Ztunnel as the mTLS agent would be deployed in its own network namespace and would require the eBPF component implement an L2 redirection mechanism.
+
+#### eBPF Component
+
+The exact function of the transparent redirection provided by the eBPF component is dependent on the networking namespace of the mTLS agent (own network namespace or root network namespace). If mTLS agent is its own network namespace, the eBPF component will redirect relevant packets from workloads and mTLS agents via L2, MAC address and interface rewrites to the node’s mTLS agent. The kernel programs could utilize a BPF map populated by the Cilium agent to determine what traffic should be redirected and to which interface.
+
+```c
+struct local_redirect_key {
+	__u64	id; // Pod IP for workloads and special well-known key value for mTLS agent
+};
+
+struct local_redirect_info {
```

Comment:
why do we need ifIndex and mac for other pods? based on my understanding, its used only for mtls pods. so instead of keeping this for every pod, should we have 2 maps. one for just storing pod ips that enforce mtls and second for mtls ip -> mac address, ifindex